src/blog/articles/node-js-race-conditions/index.md

```diff
@@ -0,0 +1,351 @@
+  await randomDelay()
+  return balance
+}
+
+async function saveBalance (value) {
+  // simulates random delay to write the data to the data storage
+  await randomDelay()
+  balance = value
+}
+
+async function sellGrapes () {
+  const balance = await loadBalance()
+  console.log(`sellGrapes - balance loaded: ${balance}`)
+  const newBalance = balance + 50
+  await saveBalance(newBalance)
+  console.log(`sellGrapes - balance updated: ${newBalance}`)
+}
+
+async function sellOlives () {
+  const balance = await loadBalance()
+  console.log(`sellOlives - balance loaded: ${balance}`)
+  const newBalance = balance + 50
+  await saveBalance(newBalance)
+  console.log(`sellOlives - balance updated: ${newBalance}`)
+}
+
+async function main () {
+  const transaction1 = sellGrapes() // NOTE: no `await`
+  const transaction2 = sellOlives() // NOTE: no `await`
+  await transaction1 // NOTE: awaiting here does not stop `transaction2` 
+                     // from being scheduled before transaction 1 is completed
+  await transaction2
+  const balance = await loadBalance()
+  console.log(`Final balance: ${balance}`)
+}
+
+main()
+```
+
+If we execute this code we might end up with different results. In one case we might get the correct outcome:
+
+```text
+sellOlives - balance loaded: 0
+sellOlives - balance updated: 50
+sellGrapes - balance loaded: 50
+sellGrapes - balance updated: 100
+Final balance: 100
+```
+
+But in other cases we might end up in a bad state:
+
+```text
+sellGrapes - balance loaded: 0
+sellOlives - balance loaded: 0
+sellGrapes - balance updated: 50
+sellOlives - balance updated: 50
+Final balance: 50
+```
+
+Note how in this last case, `sellOlives` is essentially a stale read and therefore it will end up overriding the balance disregarding any work already done by `sellGrapes`. Yes, we do have a race condition, unfortunately!
+
+Now, this example is simple ad it is not too hard to pinpoint exactly where the race condition is originated by just looking at the code.
+
+Take a minute or two to read the code again. Check out the output from the 2 cases as well. Pay attention to the notes and the log messages and try to imagine how the Node.js runtime might execute this code in the 2 different scenarios.
+
+Ok, now that you have done that, let's discuss together what happens.
+
+In our `main` function, when we execute `sellGrapes()` and `sellOlives()`, since we are not awaiting the two operations independently, we are essentially scheduling both operations onto the event loop.
+
+We only await the two transactions after they have been already scheduled, which means that they will work concurrently. After the two transactions have been scheduled, we wait for `transaction1` to finish and only then we wait for `transaction2` to complete. Note that `transaction2` might complete even before `transaction1`. In other words, awaiting for `transaction1` doesn't block `transaction2` in any way.
```

Comment:
Update this paragraph if using Promise.all([]) above. It probably gives less space for confusion and simplify this discussion a lot.